### ISSUE TITLE
[BUG] - Python scores can be arbitrarily overwritten

### ISSUE THREAD

**Issue by williamsnell** [LeetCode-Feedback/LeetCode-Feedback#25646]
### LeetCode Username

ws111

### Problem Number, Title, and Link

Every Problem | Specific example: 11. Container With Most Water, https://leetcode.com/problems/container-with-most-water/submissions/1467787998/

### Bug Category

Incorrect test case *(Output of test case is incorrect as per the problem statement)*

### Bug Description

Spotted this in the wild - using python's import machinery, the `display_runtime.txt` file can be overwritten, and arbitrary results written to the scoreboard (in this example, I have set runtime to 999999).

### Language Used for Code

Python/Python3

### Code used for Submit/Run operation

```bash
# LeetCode if you see this, you have a bug
__import__("atexit").register(lambda: open("display_runtime.txt", "w").write("999999"))

class Solution:
    def maxArea(self, height: List[int]) -> int:
        l = 0
        r = len(height) - 1

        biggest_volume = 0

        while l < r:
            height_l = height[l]
            height_r = height[r]
            biggest_volume = max(biggest_volume, min(height_l, height_r) * (r - l))
            if height_l < height_r:
                l += 1
            else:
                r -= 1

        return biggest_volume
```

### Expected behavior

A user's code should not be able to overwrite the actual calculated score.

### Screenshots

![Screenshot from 2024-12-02 11-04-22](https://github.com/user-attachments/assets/9577c3b9-d75b-4610-867c-3870005fd82a)

**Comment by bramar2** [comment]
Seems like this is also the case for C++ if you use a custom main function.

https://leetcode.com/problems/smallest-number-with-all-set-bits/submissions/1468158421/

EDIT: just found that this is possible even without a custom main function using std::atexit
https://leetcode.com/problems/smallest-number-with-all-set-bits/submissions/1468164921/

**Comment by williamsnell** [comment]
Bumping this issue since it has fairly major repercussions for the scoring system. If there's somewhere more appropriate to report this, I'm happy to move it. Thanks!

**Comment by bramar2** [comment]
Bump, a contest is coming tomorrow. This has to be fixed before then.

**Comment by extinctsion** [comment]
What! I was simplying trying to solve 4 sum issue when I saw a solution with 0ms. It was suspicious at first then I came across 1st line saying `__import__("atexit").register(lambda: open("display_runtime.txt", "w").write("0"))`. This code overwrote the runtime sec and set it to 0. You gotta fix this bug!

**Comment by takasoft** [comment]
+1 this is still a legit issue.

**Comment by takasoft** [comment]
+1 this is still a legit bug.

**Comment by sanjacob** [comment]
Average leetcode tryhard shenanigans

**Comment by sanjacob** [comment]
![Image](https://github.com/user-attachments/assets/42f08a4d-f918-4ca0-80e6-8b2c54a8b7ca)

**Comment by Ynjxsjmh** [comment]
> we have decided not to allocate development resources to fixing it at this time.

If you're not prepared to fix this issue, at the very least, you can avoid selecting these cheating results as code sample.

**Comment by ziyuang** [comment]
How did people find out the file name?

**Comment by JimouChen** [comment]
nice bug

**Comment by createyourpersonalaccount** [comment]
@ziyuang When you "Run" the code you can see the printed output, LeetCode includes it so people can debug their code, but you can use it to introspect the environment, for example:

```python
print(os.listdir(".")) # shows display_runtime.txt
```

Because LeetCode is using Python timing utilities to time the solutions, it's impossible to safeguard against this issue, as users can continue figuring out how to "disassemble" the Python environment. If they change the way they benchmark solutions, e.g. with perf, it will invalidate all the previous data they have collected from submissions, which is not fair to them.

**Comment by ziyuang** [comment]
> [@ziyuang](https://github.com/ziyuang) When you "Run" the code you can see the printed output, LeetCode includes it so people can debug their code, but you can use it to introspect the environment, for example:
> 
> print(os.listdir(".")) # shows display_runtime.txt
> Because LeetCode is using Python timing utilities to time the solutions, it's impossible to safeguard against this issue, as users can continue figuring out how to "disassemble" the Python environment. If they change the way they benchmark solutions, e.g. with perf, it will invalidate all the previous data they have collected from submissions, which is not fair to them.

I am thinking about using seccomp(-bpf) to block the `write()` system call in the user's Python process.

**Comment by createyourpersonalaccount** [comment]
> > [@ziyuang](https://github.com/ziyuang)
> I am thinking about using seccomp(-bpf) to block the `write()` system call in the user's Python process.

The LeetCode timers are written from within the Python process.

**Comment by LuaanNguyen** [comment]
Will all the submission with that line of code get removed?

**Comment by createyourpersonalaccount** [comment]
> Will all the submission with that line of code get removed?

No, you can always obfuscate it. Just accept that LeetCode is not going to fix this.

**Comment by benjaminwp18** [comment]
Guess I'm late to the party—I noticed today the best runtime code sample on `11. Container With Most Water` still uses this bug. They obfuscate their solution with `exec("\x5F\x5F\x69\x6D\x70\x6F...")`. The unobfuscated code writes 0 to `display_runtime.txt`, then writes a hardcoded string of problem solutions separated by newlines to `user.out`:
```python
open('user.out', 'w').write(
"""49
1
16
2
1
1
...
""")
```

Not sure how well that agrees with the LeetCode Support Team response:

> While it is technically possible for a user to alter their own time display, this does not impact rankings, contest fairness, or *any other users’ experience*.

Unlikely LC will read this, but if proper sandboxing isn't an option...
  - randomizing test order and
  - overwriting the atexit module (or the import function) to keep people from getting hooks after `display_runtime.txt` is initially written

...might help?

**Comment by shaishab316** [comment]
<img width="1912" height="910" alt="Image" src="https://github.com/user-attachments/assets/506ef9e2-284b-46ad-8b21-2abdc0e17a3e" />

Nothing to say 😴 0ms
